plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt

```diff
@@ -961,21 +977,32 @@ class JavaToJKTreeBuilder(
             }
         }
 
-        fun PsiCodeBlock.toJK(): JKBlock = JKBlockImpl(
-            if (withBody) statements.map { it.toJK() } else listOf(createTodoExpression().asStatement())
-        ).withFormattingFrom(this).also {
-            it.leftBrace.withFormattingFrom(lBrace)
-            it.rightBrace.withFormattingFrom(rBrace)
+        fun PsiCodeBlock.toJK(): JKBlock {
+            if (!immutableVariables.keys.contains(this)) {
+                immutableVariables[this] = LocalCanBeFinal().getImmutableLocalVariablesInBlock(this)
+            }
+            return JKBlockImpl(
+                if (withBody) statements.map { it.toJK() } else listOf(createTodoExpression().asStatement())
+            ).withFormattingFrom(this).also {
+                it.leftBrace.withFormattingFrom(lBrace)
+                it.rightBrace.withFormattingFrom(rBrace)
+            }
         }
 
         fun PsiLocalVariable.toJK(): JKLocalVariable {
+            val codeBlock = parentOfType<PsiCodeBlock>()
+            val mutability = when {
+                hasModifierProperty(PsiModifier.FINAL) -> IMMUTABLE
+                immutableVariables[codeBlock]?.any { it == this } == true -> IMMUTABLE
```

Comment:
`immutableVariables[codeBlock]` is a `Set`. `immutableVariables[codeBlock]?.contains(this) == true` should be faster :)